op-node/cmd/p2p/req_enr.go

```diff
@@ -0,0 +1,206 @@
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/p2p/enr"
+	"github.com/ethereum/go-ethereum/rlp"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/metrics"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	rollupP2P "github.com/ethereum-optimism/optimism/op-node/p2p"
+	rollupCli "github.com/ethereum-optimism/optimism/op-node/p2p/cli"
+	rollup "github.com/ethereum-optimism/optimism/op-node/rollup"
+)
+
+// enrPrefix is the prefix for enrs.
+const enrPrefix = "enr:"
+
+// b64format is the base64 encoding format used for enrs.
+var b64format = base64.RawURLEncoding
+
+// RequestEnr requests an enr from a node peer.
+func RequestEnr(ctx *cli.Context) error {
+	quitC := make(chan struct{})
+
+	log.Root().SetHandler(
+		log.LvlFilterHandler(log.LvlDebug, log.StreamHandler(os.Stdout, log.TerminalFormat(true))),
+	)
+
+	// Read the enr from the command line
+	enrStr := ctx.String("enr")
+	if enrStr == "" {
+		enrStr = ctx.Args().First()
+	}
+	if enrStr == "" {
+		return fmt.Errorf("no enr provided")
+	}
+	log.Debug("Found enr", "enr", enrStr)
+
+	// Read in the block time
+	blockTime := ctx.Uint64("block-time")
+	if blockTime == 0 {
+		blockTime = 2
+		log.Info("No block time provided, using default", "block-time", blockTime)
+	}
+
+	// Parse the enr argument string
+	remoteNode, err := ParseENR(enrStr, enode.ValidSchemes)
+	if err != nil {
+		return fmt.Errorf("failed to parse enr: %w", err)
+	}
+	log.Debug("Parsed enr", "enr", remoteNode.String())
+
+	// Create a new op-node rollup p2p config
+	p2pConfig, err := rollupCli.NewConfig(ctx, blockTime)
+	if err != nil {
+		return fmt.Errorf("failed to load p2p config: %w", err)
+	}
+	log.Debug("Created p2p config, starting discovery...")
+
+	// Create a new host
+	bwc := metrics.NewBandwidthCounter()
+	host, err := p2pConfig.Host(log.Root(), bwc)
+	if err != nil {
+		return fmt.Errorf("failed to create host: %w", err)
+	}
+	log.Info("Created host", "addrs", host.Addrs(), "peerID", host.ID().Pretty())
+
+	// Monitor peers in a goroutine
+	log.Info("Monitoring peers...")
+	go MonitorPeers(host, log.Root(), quitC)
+
+	// Find the active p2p port
+	tcpPort, err := rollupP2P.FindActiveTCPPort(host)
+	if err != nil {
+		return fmt.Errorf("failed to find active tcp port: %w", err)
+	}
+	log.Debug("Found active tcp port", "tcpPort", tcpPort)
+
+	// Notify of any new connections/streams/etc.
+	host.Network().Notify(rollupP2P.NewNetworkNotifier(log.Root(), nil))
+	log.Info("Setup new network notifier")
+
+	// All nil if disabled.
+	rollupCfg := &rollup.Config{
+		L2ChainID: big.NewInt(100),
+	}
+	_, dv5Udp, err := p2pConfig.Discovery(log.New("p2p", "discv5"), rollupCfg, tcpPort)
+	if err != nil {
+		return fmt.Errorf("failed to start discv5: %w", err)
+	}
+
+	log.Debug("Discv5 started, requesting enr...")
+
+	for {
```

Comment:
this entire `for` loop could go in a goroutine, then we could block like this
https://github.com/ethereum-optimism/optimism/blob/fc56ded13c91ae945f7e35bff0e45f9154fd647e/op-batcher/batcher/batch_submitter.go#L102